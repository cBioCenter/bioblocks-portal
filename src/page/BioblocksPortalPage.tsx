import { EMPTY_FUNCTION, fetchDataset } from 'bioblocks-viz';
import * as React from 'react';
import { connect } from 'react-redux';
import { Route, RouteComponentProps, Switch } from 'react-router-dom';
import { bindActionCreators, Dispatch } from 'redux';
import { Container } from 'semantic-ui-react';

import { ConnectedRouter } from 'connected-react-router';
import { SiteHeader } from '~bioblocks-portal~/container';
import { IDataset, IEveResponse, IVignette, IVisualization } from '~bioblocks-portal~/data';
import {
  DatasetsPage,
  DynamicsPage,
  LandingPage,
  VignettesPage,
  VisualizationsPage,
  VizOverviewPage,
} from '~bioblocks-portal~/page';
import { history, IPortalReducerState } from '~bioblocks-portal~/reducer';
import { selectDatasets, selectVignettes, selectVisualizations } from '~bioblocks-portal~/selector';

export interface IBioblocksPortalPageProps {
  datasets: IDataset[];
  vignettes: IVignette[];
  visualizations: IVisualization[];
  dispatchDatasetsFetch(dataset: string, fetchFn: () => Promise<IDataset[]>): void;
  dispatchVignettesFetch(dataset: string, fetchFn: () => Promise<IVignette[]>): void;
  dispatchVisualizationsFetch(dataset: string, fetchFn: () => Promise<IVisualization[]>): void;
}

export class UnconnectedBioblocksPortalPage extends React.Component<IBioblocksPortalPageProps> {
  public static defaultProps = {
    datasets: new Array<IDataset>(),
    dispatchDatasetsFetch: EMPTY_FUNCTION,
    dispatchVignettesFetch: EMPTY_FUNCTION,
    dispatchVisualizationsFetch: EMPTY_FUNCTION,
    vignettes: new Array<IVignette>(),
    visualizations: new Array<IVisualization>(),
  };

  constructor(props: IBioblocksPortalPageProps) {
    super(props);
  }

  public async componentDidMount() {
    const { dispatchDatasetsFetch, dispatchVignettesFetch, dispatchVisualizationsFetch } = this.props;

    dispatchDatasetsFetch('datasets', this.onDatasetFetch);

    dispatchVignettesFetch('vignettes', this.onVignetteFetch);

    dispatchVisualizationsFetch('visualizations', this.onVisualizationFetch);
  }

  public render() {
    return (
      <ConnectedRouter history={history}>
        <Container id={'BioblocksPortalPage'} fluid={true}>
          <SiteHeader {...this.props} />

          <Switch>
            <Route exact={true} strict={true} path={'/datasets'} render={this.renderDatasetsPage} />
            <Route exact={true} strict={true} path={'/visualizations'} render={this.renderVisualizationsPage} />
            <Route exact={true} strict={true} path={'/visualizations/'} render={this.renderVizOverviewPage} />
            <Route path={'/dynamics'} render={this.renderDynamicsPage} />
            <Route path={'/vignettes'} render={this.renderVignettesPage} />
            <Route exact={true} path={'/'} render={this.renderLandingPage} />
          </Switch>
        </Container>
      </ConnectedRouter>
    );
  }

  protected onDatasetFetch = async () => {
    const datasetFetchResult = await fetch(`${process.env.API_URL}/dataset?embedded={"analyses": 1}`);
    if (!datasetFetchResult.ok) {
      return [];
    } else {
      const response = (await datasetFetchResult.json()) as IEveResponse<IDataset>;

      return response._items.sort((a, b) => a.name.localeCompare(b.name));
    }
  };

  protected onVignetteFetch = async () => {
    const vignetteFetchResult = await fetch(`${process.env.API_URL}/vignette`);
    if (!vignetteFetchResult.ok) {
      return [];
    } else {
      const response = (await vignetteFetchResult.json()) as IEveResponse<IVignette>;

      return response._items.sort((a, b) => a.name.localeCompare(b.name));
    }
  };

  protected onVisualizationFetch = async () => {
    const visualizationFetchResult = await fetch(`${process.env.API_URL}/visualization`);
    if (!visualizationFetchResult.ok) {
      return [];
    } else {
      const response = (await visualizationFetchResult.json()) as IEveResponse<IVisualization>;

      return response._items.sort((a, b) => a.name.localeCompare(b.name));
    }
  };

  protected renderDynamicsPage = () => {
    return <DynamicsPage />;
  };

  protected renderLandingPage = (props: RouteComponentProps) => {
    const { vignettes, visualizations } = this.props;

    const featuredVignettes = new Array<IVignette>();
    if (vignettes) {
      const tabulaVignette = vignettes.find(vignette => vignette._id === 'bbdecade-0000-0000-0001-a1234567890b');
      const springVignette = vignettes.find(vignette => vignette._id === 'bbdecade-0000-0000-0002-a1234567890b');

      if (springVignette) {
        featuredVignettes.push(springVignette);
      }
      if (tabulaVignette) {
        featuredVignettes.push(tabulaVignette);
      }
    }

    const featuredVisualizations = new Array<IVisualization>();
    if (visualizations) {
      const springViz = visualizations.find(viz => viz.name === 'SPRING');
      const umapViz = visualizations.find(viz => viz.name === 'UMAP');

      if (springViz) {
        featuredVisualizations.push(springViz);
      }
      if (umapViz) {
        featuredVisualizations.push(umapViz);
      }
    }

    return (
      <LandingPage {...props} featuredVignettes={featuredVignettes} featuredVisualizations={featuredVisualizations} />
    );
  };

  protected renderDatasetsPage = (props: RouteComponentProps) => {
    const { datasets, visualizations } = this.props;

    return <DatasetsPage datasets={datasets} visualizations={visualizations} {...props} />;
  };

  protected renderVignettesPage = (props: RouteComponentProps) => {
    const { vignettes } = this.props;

    return <VignettesPage {...props} vignettes={vignettes} />;
  };

  protected renderVisualizationsPage = (props: RouteComponentProps) => {
    const { visualizations } = this.props;

    return <VisualizationsPage {...props} visualizations={visualizations} />;
  };

  protected renderVizOverviewPage = (props: RouteComponentProps) => {
    const { vignettes, visualizations } = this.props;

    return <VizOverviewPage {...props} vignettes={vignettes} visualizations={visualizations} />;
  };
}

const mapStateToProps = (state: IPortalReducerState) => ({
  datasets: selectDatasets(state),
  vignettes: selectVignettes(state),
  visualizations: selectVisualizations(state),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      dispatchDatasetsFetch: fetchDataset,
      dispatchVignettesFetch: fetchDataset,
      dispatchVisualizationsFetch: fetchDataset,
    },
    dispatch,
  );

export const BioblocksPortalPage = connect(
  mapStateToProps,
  mapDispatchToProps,
)(UnconnectedBioblocksPortalPage);
