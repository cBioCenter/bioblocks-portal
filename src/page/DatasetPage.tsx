import * as React from 'react';
import { Grid, Message } from 'semantic-ui-react';

import { AnatomogramContainer, ISpringGraphData, SPECIES_TYPE, SpringContainer, TensorTContainer } from 'bioblocks-viz';
import { connect } from 'react-redux';
import { IVignette, IVisualization } from '~bioblocks-portal~/data';
import { IPortalReducerState } from '~bioblocks-portal~/reducer';
import { selectVignettes, selectVisualizations } from '~bioblocks-portal~/selector';

export interface IDatasetPageProps {
  pathname: string;
  search: string;
  vignettes: IVignette[];
  visualizations: IVisualization[];
  dispatchSpringFetch(fetchFn: () => Promise<ISpringGraphData>, namespace?: string): void;
  setSpecies(species: SPECIES_TYPE): void;
}

export interface IDatasetPageState {
  datasetVisualizations: IVisualization[];
  datasetLocation: string;
}

export class UnconnectedDatasetPage extends React.Component<IDatasetPageProps, IDatasetPageState> {
  public static defaultProps = {
    dispatchSpringFetch: () => {
      return;
    },
    setSpecies: () => {
      return;
    },
  };

  constructor(props: IDatasetPageProps) {
    super(props);
    this.state = {
      datasetLocation: 'hpc_sf2/full',
      datasetVisualizations: [],
    };
  }

  public async componentDidMount() {
    const { search, setSpecies } = this.props;
    const { datasetLocation } = this.state;
    if (search) {
      this.setupSearchParameters(search);
    }
    setSpecies(datasetLocation.includes('hpc') ? 'homo_sapiens' : 'mus_musculus');
  }

  public componentDidUpdate(prevProps: IDatasetPageProps) {
    const { search, setSpecies, vignettes, visualizations } = this.props;
    const { datasetLocation } = this.state;
    if (
      (search && search !== prevProps.search) ||
      vignettes !== prevProps.vignettes ||
      visualizations !== prevProps.visualizations
    ) {
      this.setupSearchParameters(search);
    }
    setSpecies(datasetLocation.includes('hpc') ? 'homo_sapiens' : 'mus_musculus');
  }

  public render() {
    const { datasetVisualizations, datasetLocation } = this.state;

    return (
      <div style={{ padding: '20px' }}>
        <Grid centered={true} stackable={true} stretched={false} padded={true} columns={2}>
          {datasetLocation.length >= 1 &&
            datasetVisualizations.map((visualization, index) => (
              <Grid.Column key={`dataset-visualization-${index}`} style={{ width: 'auto' }}>
                {this.renderVisualization(visualization, datasetLocation)}
              </Grid.Column>
            ))}
        </Grid>
      </div>
    );
  }

  protected setupSearchParameters(query: string) {
    const { vignettes, visualizations } = this.props;

    let datasetLocation = this.state.datasetLocation;
    const datasetVisualizations = new Array<IVisualization>();
    if (vignettes) {
      const vignetteParams = new URLSearchParams(query);
      // tslint:disable-next-line:no-backbone-get-set-outside-model
      const datasetId = vignetteParams.get('id');

      datasetLocation = datasetId ? datasetId : datasetLocation;
      const vizIds = vignetteParams.getAll('viz');
      vizIds.forEach(vizId => {
        const viz = visualizations.find(aViz => vizId === aViz._id);
        if (viz) {
          datasetVisualizations.push(viz);
        }
      });

      this.setState({
        datasetLocation,
        datasetVisualizations,
      });
    }
  }

  protected renderVisualization(viz: IVisualization, datasetLocation: string) {
    const isFullPage = this.state.datasetVisualizations.length === 1;
    const iconSrc = `${process.env.API_URL}${viz.icon}`;
    console.log(datasetLocation);
    switch (viz.name.toLocaleLowerCase()) {
      case 'spring':
        return (
          <SpringContainer
            datasetLocation={datasetLocation}
            datasetsURI={`${process.env.API_URL}/datasets`}
            isFullPage={isFullPage}
            springSrc={`${process.env.API_URL}/${viz.location}`}
            iconSrc={iconSrc}
          />
        );
      case 'tfjs-tsne':
        return (
          <TensorTContainer
            datasetLocation={`${process.env.API_URL}/datasets/${datasetLocation}`}
            iconSrc={iconSrc}
            isFullPage={isFullPage}
          />
        );
      case 'anatomogram':
        return <AnatomogramContainer iconSrc={iconSrc} />;
      default:
        return <Message error={true}>{`Currently unsupported visualization '${viz}'`}</Message>;
    }
  }
}

const mapStateToProps = (state: IPortalReducerState) => ({
  pathname: state.router.location.pathname,
  search: state.router.location.search,
  vignettes: selectVignettes(state),
  visualizations: selectVisualizations(state),
});

export const DatasetPage = connect(mapStateToProps)(UnconnectedDatasetPage);
