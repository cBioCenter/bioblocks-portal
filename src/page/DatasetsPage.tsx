// ~bioblocks-portal~
// Datasets Page
// Page for displaying datasets available on bioblocks-server.
// ~bioblocks-portal~

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Button, Divider, Grid, Header } from 'semantic-ui-react';

import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { IAnalysis, IDataset, IVisualization } from '~bioblocks-portal~/data';

export interface IDatasetsPageProps extends Partial<RouteComponentProps> {
  datasets: IDataset[];
  visualizations: IVisualization[];
}

export class UnconnectedDatasetsPage extends React.Component<IDatasetsPageProps, any> {
  public static defaultProps = {
    datasets: [],
    visualizations: [],
  };

  constructor(props: IDatasetsPageProps) {
    super(props);
  }

  public render() {
    const { datasets } = this.props;

    return (
      <Grid centered={true} padded={true} relaxed={true}>
        {datasets
          .sort((a, b) => b.analyses.length - a.analyses.length)
          .map((dataset, index) => (
            <React.Fragment key={`vignette-${index}`}>{this.renderSingleDatasetEntry(dataset)}</React.Fragment>
          ))}
      </Grid>
    );
  }

  protected getValidVizIds = (analyses: IAnalysis[]) => {
    let result = '';

    for (const analysis of analyses) {
      if (analysis.processType === 'SPRING') {
        result += '&viz=bbfacade-0000-0000-0001-a1234567890b';
        result += '&viz=bbfacade-0000-0000-0004-a1234567890b';
      } else if (analysis.processType === 'TSNE') {
        // result += '&viz=bbfacade-0000-0000-0003-a1234567890b';
      } else if (analysis.processType === 'UMAP') {
        result += '&viz=bbfacade-0000-0000-0004-a1234567890b';
      } else {
        console.log(`Process ${analysis._id} has unhandled process of type '${analysis.processType}'.`);
      }
    }

    return result;
  };

  protected renderSingleDatasetEntry(dataset: IDataset) {
    return (
      <>
        <Grid.Row columns={3}>{this.renderSingleDatasetEntryHeader(dataset)}</Grid.Row>
        <Divider section={true} />
      </>
    );
  }

  protected renderDatasetAnalysisLink(analysis: IAnalysis, datasetId: string) {
    const linkTo = `/dynamics?id=${datasetId}&analysis=${analysis._id}${this.getValidVizIds([analysis])}`;

    return (
      <Button basic={true} key={`launch-button-for-${analysis._id}`}>
        <Link to={linkTo}>{`Launch '${analysis.name}'`}</Link>
      </Button>
    );
  }

  protected renderSingleDatasetEntryHeader(dataset: IDataset) {
    return (
      <Grid.Column textAlign={'left'}>
        <Header>{dataset.name}</Header>

        <Grid.Row>{dataset.analyses.map(analysis => this.renderDatasetAnalysisLink(analysis, dataset._id))}</Grid.Row>
      </Grid.Column>
    );
  }
}

export const DatasetsPage = connect(undefined)(UnconnectedDatasetsPage);
