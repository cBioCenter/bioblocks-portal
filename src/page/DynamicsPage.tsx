import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';
import { Grid, Message } from 'semantic-ui-react';

import {
  AnatomogramContainer,
  EMPTY_FUNCTION,
  fetchDataset,
  fetchJSONFile,
  fetchMatrixData,
  ICategoricalAnnotation,
  SPECIES_TYPE,
  SpringContainer,
  TensorTContainer,
  UMAPTranscriptionalContainer,
} from 'bioblocks-viz';

import { IDataset, IVignette, IVisualization } from '~bioblocks-portal~/data';
import { IPortalReducerState } from '~bioblocks-portal~/reducer';
import { selectVignettes, selectVisualizations } from '~bioblocks-portal~/selector';

export interface IDynamicsPageProps {
  dataset: IDataset | null;
  pathname: string;
  search: string;
  vignettes: IVignette[];
  visualizations: IVisualization[];
  dispatchDatasetFetch(dataset: string, fetchFn: () => Promise<IDataset | null>): void;
}

export interface IDynamicsPageState {
  datasetVisualizations: IVisualization[];
  datasetLocation: string;
  scRNAseqCategorySelected: string;
  scRNAseqCategoricalData: ICategoricalAnnotation;
  scRNAseqMatrix: number[][];
}

export class UnconnectedDynamicsPage extends React.Component<IDynamicsPageProps, IDynamicsPageState> {
  public static defaultProps = {
    dataset: null,
    dispatchDatasetFetch: EMPTY_FUNCTION,
    pathname: '',
    search: '',
    vignettes: new Array<IVignette>(),
    visualizations: new Array<IVisualization>(),
  };

  constructor(props: IDynamicsPageProps) {
    super(props);
    this.state = {
      datasetLocation: 'hpc_sf2/full',
      datasetVisualizations: [],
      scRNAseqCategoricalData: {},
      scRNAseqCategorySelected: 'Sample',
      scRNAseqMatrix: new Array(new Array<number>()),
    };
  }

  public async componentDidMount() {
    const { search } = this.props;
    if (search) {
      await this.setupSearchParameters(search);
    }
  }

  public async componentDidUpdate(prevProps: IDynamicsPageProps) {
    const { search, vignettes, visualizations } = this.props;
    if (
      (search && search !== prevProps.search) ||
      vignettes !== prevProps.vignettes ||
      visualizations !== prevProps.visualizations
    ) {
      await this.setupSearchParameters(search);
    }
  }

  public render() {
    const { datasetVisualizations, datasetLocation } = this.state;

    return (
      <div style={{ padding: '20px' }}>
        <Grid centered={true} stackable={true} stretched={false} padded={true} columns={2}>
          {datasetVisualizations.length >= 1 ? (
            datasetLocation.length >= 1 &&
            datasetVisualizations.map((visualization, index) => (
              <Grid.Column key={`dataset-visualization-${index}`} style={{ width: 'auto' }}>
                {this.renderVisualization(visualization, datasetLocation)}
              </Grid.Column>
            ))
          ) : (
            <Grid.Column>No visualizations!</Grid.Column>
          )}
        </Grid>
      </div>
    );
  }

  protected async setupSearchParameters(query: string) {
    const { dispatchDatasetFetch, vignettes, visualizations } = this.props;

    let datasetLocation = this.state.datasetLocation;
    const datasetVisualizations = new Array<IVisualization>();
    if (vignettes) {
      const vignetteParams = new URLSearchParams(query);
      // tslint:disable-next-line:no-backbone-get-set-outside-model
      const datasetId = vignetteParams.get('id');

      datasetLocation = datasetId ? datasetId : datasetLocation;
      const vizIds = vignetteParams.getAll('viz');
      vizIds.forEach(vizId => {
        const viz = visualizations.find(aViz => vizId === aViz._id);
        if (viz) {
          datasetVisualizations.push(viz);
        }
      });
      dispatchDatasetFetch('dataset', async () => {
        const datasetFetchResult = await fetch(`${process.env.API_URL}/dataset/${datasetId}?embedded={"analyses":1}`);
        if (!datasetFetchResult.ok) {
          return null;
        } else {
          return (await datasetFetchResult.json()) as IDataset;
        }
      });

      let scRNAseqCategoricalData = this.state.scRNAseqCategoricalData;
      let scRNAseqMatrix = this.state.scRNAseqMatrix;
      let scRNAseqCategorySelected = this.state.scRNAseqCategorySelected;
      if (this.props.dataset) {
        const springAnalysis = this.props.dataset.analyses.find(anAnalysis => anAnalysis.processType === 'SPRING');
        if (springAnalysis) {
          const springLocation = `${process.env.API_URL}/datasets/${datasetLocation}/analyses/${springAnalysis._id}`;
          scRNAseqCategoricalData = (await fetchJSONFile(
            `${springLocation}/${this.props.dataset.name}/categorical_coloring_data.json`,
          )) as ICategoricalAnnotation;
          scRNAseqMatrix = await fetchMatrixData(`${springLocation}/pca.csv`);
          scRNAseqCategorySelected =
            Object.keys(scRNAseqCategoricalData).length >= 1
              ? Object.keys(scRNAseqCategoricalData)[0]
              : scRNAseqCategorySelected;
        }
      }

      this.setState({
        datasetLocation,
        datasetVisualizations,
        scRNAseqCategoricalData,
        scRNAseqCategorySelected,
        scRNAseqMatrix,
      });
    }
  }

  protected renderSpringVisualization(
    viz: IVisualization,
    isFullPage: boolean,
    datasetLocation: string,
    iconSrc: string,
  ) {
    let analysisLocation = '';
    if (this.props.dataset) {
      const analysis = this.props.dataset.analyses.find(anAnalysis => anAnalysis.processType === 'SPRING');
      if (analysis) {
        analysisLocation = analysis._id;
      }
    }

    const datasetName = this.props.dataset ? this.props.dataset.name : '';

    return (
      <SpringContainer
        datasetLocation={`${datasetLocation}/analyses/${analysisLocation}/${datasetName}`}
        datasetsURI={`${process.env.API_URL}/datasets`}
        isFullPage={isFullPage}
        springSrc={`${process.env.API_URL}/${viz.location}`}
        iconSrc={iconSrc}
      />
    );
  }

  protected renderUMAPVisualization() {
    const { scRNAseqCategoricalData, scRNAseqMatrix, scRNAseqCategorySelected } = this.state;

    if (this.props.dataset) {
      return (
        <UMAPTranscriptionalContainer
          categoricalAnnotations={scRNAseqCategoricalData}
          dataMatrix={scRNAseqMatrix}
          labelCategory={scRNAseqCategorySelected}
          numIterationsBeforeReRender={1}
          numSamplesToShow={5000}
          sampleNames={undefined}
        />
      );
    } else {
      return null;
    }
  }

  protected renderVisualization(viz: IVisualization, datasetLocation: string) {
    const isFullPage = this.state.datasetVisualizations.length === 1;
    const iconSrc = `${process.env.API_URL}${viz.icon}`;
    let analysisLocation = '';
    switch (viz.name.toLocaleLowerCase()) {
      case 'spring':
        return this.renderSpringVisualization(viz, isFullPage, datasetLocation, iconSrc);
      case 'tfjs-tsne':
        if (this.props.dataset) {
          const tsneAnalysis = this.props.dataset.analyses.find(analysis => analysis.processType === 'TSNE');
          if (tsneAnalysis) {
            analysisLocation = tsneAnalysis._id;
          }
        }

        return (
          <TensorTContainer
            datasetLocation={`${process.env.API_URL}/datasets/${datasetLocation}/analyses/${analysisLocation}`}
            iconSrc={iconSrc}
          />
        );
      case 'anatomogram':
        const species: SPECIES_TYPE =
          this.props.dataset && this.props.dataset.species === 'mus_musculus' ? 'mus_musculus' : 'homo_sapiens';

        return <AnatomogramContainer species={species} iconSrc={iconSrc} />;
      case 'umap':
        return this.renderUMAPVisualization();

      default:
        return <Message error={true}>{`Currently unsupported visualization '${viz}'`}</Message>;
    }
  }
}

const mapStateToProps = (state: IPortalReducerState) => ({
  dataset: state.dataset,
  pathname: state.router.location.pathname,
  search: state.router.location.search,
  vignettes: selectVignettes(state),
  visualizations: selectVisualizations(state),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
  bindActionCreators(
    {
      dispatchDatasetFetch: fetchDataset as ((dataset: string, fetchFn: () => Promise<IDataset | null>) => void),
    },
    dispatch,
  );

export const DynamicsPage = connect(
  mapStateToProps,
  mapDispatchToProps,
)(UnconnectedDynamicsPage);
