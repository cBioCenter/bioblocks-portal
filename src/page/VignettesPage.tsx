// ~bioblocks-portal~
// Vignettes Page
// Page to display information about vignettes from bioblocks-server.
// ~bioblocks-portal~

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { Link } from 'react-router-dom';
import { Button, Grid, Header } from 'semantic-ui-react';

import { connect } from 'react-redux';
import { getFormattedAuthors, IVignette } from '~bioblocks-portal~/data';

export interface IVignettesPageProps extends Partial<RouteComponentProps> {
  vignettes: IVignette[];
}

export class UnconnectedVignettesPage extends React.Component<IVignettesPageProps, any> {
  public static defaultProps = {
    vignettes: [],
  };

  constructor(props: IVignettesPageProps) {
    super(props);
  }

  public render() {
    const { vignettes } = this.props;

    return (
      <Grid centered={true} padded={true} relaxed={true}>
        {vignettes.map((vignette, index) => (
          <React.Fragment key={`vignette-${index}`}>{this.renderSingleVignette(vignette)}</React.Fragment>
        ))}
      </Grid>
    );
  }

  protected renderSingleVignette(vignette: IVignette) {
    return (
      <Grid.Row columns={3}>
        <Grid.Column>
          <img
            src={`${process.env.API_URL}${vignette.icon}`}
            style={{ height: '90px', width: '90px' }}
            alt={`vignette ${vignette.name} icon`}
          />
        </Grid.Column>
        {this.renderVignetteSummary(vignette)}
        {this.renderVignetteLink(vignette)}
      </Grid.Row>
    );
  }

  protected renderVignetteLink(vignette: IVignette) {
    return (
      <Grid.Column>
        <Button basic={true}>
          <Link to={`/dynamics?id=${vignette.dataset}${vignette.visualizations.map(viz => `&viz=${viz}`).join('')}`}>
            launch
          </Link>
        </Button>
      </Grid.Column>
    );
  }

  protected renderVignetteSummary(vignette: IVignette) {
    return (
      <Grid.Column textAlign={'left'}>
        <Header>{vignette.name}</Header>
        <p>
          <span style={{ fontWeight: 'bold' }}>Description: </span>
          {vignette.summary}
        </p>
        <p>
          <span style={{ fontWeight: 'bold' }}>Analysis authors: </span>
          {getFormattedAuthors(vignette)}
        </p>
        <br />
      </Grid.Column>
    );
  }
}

export const VignettesPage = connect(undefined)(UnconnectedVignettesPage);
