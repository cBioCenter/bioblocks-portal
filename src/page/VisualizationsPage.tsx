import * as React from "react";
import { RouteComponentProps } from "react-router";
import { Link } from "react-router-dom";
import { Button, Grid, Header, List } from "semantic-ui-react";

import { IVisualization } from "~bioblocks-portal~/data";

export interface IVisualizationsPageProps extends Partial<RouteComponentProps> {
  visualizations: IVisualization[];
}

export class VisualizationsPage extends React.Component<
  IVisualizationsPageProps,
  any
> {
  public static defaultProps = {
    visualizations: []
  };

  constructor(props: IVisualizationsPageProps) {
    super(props);
  }

  public render() {
    const { visualizations } = this.props;

    return (
      <List divided={true}>
        {visualizations.map((viz, index) => (
          <List.Item key={`viz-page-li-${index}`}>
            {this.renderVisualizationItem(viz)}
          </List.Item>
        ))}
      </List>
    );
  }

  protected renderVisualizationItem(viz: IVisualization) {
    return (
      <Grid centered={true} columns={3} padded={true} relaxed={true}>
        <Grid.Column>
          <img
            src={`${process.env.API_URL}${viz.icon}`}
            alt={`icon for ${viz.name}`}
            style={{ height: "100px" }}
          />
        </Grid.Column>
        <Grid.Column textAlign={"left"}>
          <Header as={"h2"}>{viz.name}</Header>
          <List>
            <List.Item>{`${
              viz.isOriginal ? "original: " : ""
            }${viz.authors.join(", ")}`}</List.Item>
            <List.Item>{viz.summary}</List.Item>
            <List.Item>{`relevant data: ${viz.compatibleData.join(
              ", "
            )}`}</List.Item>
          </List>
        </Grid.Column>
        <Grid.Column stretched={true}>
          <Grid.Row>
            <Button basic={true}>
              <Link
                to={{ pathname: "/visualizations/", search: `?id=${viz._id}` }}
              >
                details
              </Link>
            </Button>
          </Grid.Row>
          <Grid.Row>
            <Button basic={true}>
              <Link
                to={{
                  pathname: "/dataset",
                  search: `?id=${viz.exampleDataset}&viz=${viz._id}`
                }}
              >
                launch
              </Link>
            </Button>
          </Grid.Row>
        </Grid.Column>
      </Grid>
    );
  }
}
