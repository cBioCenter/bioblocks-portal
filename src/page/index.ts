export * from './BioblocksPortalPage';
export * from './VisualizationsPage';
export * from './DatasetPage';
export * from './LandingPage';
export * from './VignettesPage';
export * from './VizOverviewPage';
