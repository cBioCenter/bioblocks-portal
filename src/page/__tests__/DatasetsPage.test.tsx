import { ConnectedRouter } from 'connected-react-router';
import { mount, shallow } from 'enzyme';
import { createMemoryHistory } from 'history';
import * as React from 'react';
import { Provider } from 'react-redux';
import { Header } from 'semantic-ui-react';

import { IDataset } from '~bioblocks-portal~/data';
import { DatasetsPage, UnconnectedDatasetsPage } from '~bioblocks-portal~/page';
import { configureStore } from '~bioblocks-portal~/reducer';

describe('DatasetsPage', () => {
  it('Should match existing snapshot when no props are provided.', () => {
    const wrapper = shallow(<UnconnectedDatasetsPage />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should match existing snapshot when given a dataset.', () => {
    const sampleDatasets: IDataset[] = [
      {
        _id: 'some-kind-of-musical-reference',
        analyses: [],
        authors: [],
        derivedFrom: [],
        matrixLocation: '',
        name: '',
      },
    ];
    const wrapper = shallow(<UnconnectedDatasetsPage datasets={sampleDatasets} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should render dynamics links correctly.', () => {
    const sampleDatasets: IDataset[] = [
      {
        _id: 'some-kind-of-musical-reference',
        analyses: [
          {
            _id: 'down-the-road...',
            processType: 'SPRING',
          },
          {
            _id: 'and-back-again...',
            processType: 'TSNE',
          },
        ],
        authors: [],
        derivedFrom: [],
        matrixLocation: '',
        name: '',
      },
    ];
    const wrapper = shallow(<UnconnectedDatasetsPage datasets={sampleDatasets} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Should sort datasets with analysis links before those without.', () => {
    const sampleDatasets: IDataset[] = [
      {
        _id: 'blend',
        analyses: [],
        authors: [],
        derivedFrom: [],
        matrixLocation: '',
        name: 'blend',
      },
      {
        _id: 'will',
        analyses: [
          {
            _id: 'yes',
            processType: 'SPRING',
          },
          {
            _id: 'no',
            processType: 'SPRING',
          },
        ],
        authors: [],
        derivedFrom: [],
        matrixLocation: '',
        name: 'will',
      },
      {
        _id: 'it',
        analyses: [
          {
            _id: 'maybe',
            processType: 'SPRING',
          },
        ],
        authors: [],
        derivedFrom: [],
        matrixLocation: '',
        name: 'it',
      },
    ];
    const wrapper = shallow(<UnconnectedDatasetsPage datasets={sampleDatasets} />);
    const headers = wrapper.find(Header);
    expect(
      headers
        .at(0)
        .childAt(0)
        .text(),
    ).toEqual('will');
    expect(
      headers
        .at(1)
        .childAt(0)
        .text(),
    ).toEqual('it');
    expect(
      headers
        .at(2)
        .childAt(0)
        .text(),
    ).toEqual('blend');
  });

  it('Should match existing snapshot when hooked up to a Redux store.', () => {
    const store = configureStore();
    const history = createMemoryHistory();
    const wrapper = mount(
      <Provider store={store}>
        <ConnectedRouter history={history}>
          <DatasetsPage />
        </ConnectedRouter>
      </Provider>,
    );
    history.push('/dynamics?id=psycho-jukebox-vignette-id');
    wrapper.update();
    expect(wrapper.find(DatasetsPage)).toMatchSnapshot();
  });
});
