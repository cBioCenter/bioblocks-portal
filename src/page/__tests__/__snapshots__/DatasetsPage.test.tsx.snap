// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetsPage Should match existing snapshot when given a dataset. 1`] = `
<Grid
  centered={true}
  padded={true}
  relaxed={true}
>
  <GridRow
    columns={3}
  >
    <GridColumn
      textAlign="left"
    >
      <Header />
      <p>
        <span
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Analyses: 0
        </span>
      </p>
      <p>
        <span
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Derived from: 
          Nothing
        </span>
      </p>
      <br />
    </GridColumn>
    <GridColumn />
  </GridRow>
</Grid>
`;

exports[`DatasetsPage Should match existing snapshot when hooked up to a Redux store. 1`] = `
<Connect(UnconnectedDatasetsPage)>
  <UnconnectedDatasetsPage
    datasets={Array []}
    dispatch={[Function]}
    visualizations={Array []}
  >
    <Grid
      centered={true}
      padded={true}
      relaxed={true}
    >
      <div
        className="ui centered padded relaxed grid"
      />
    </Grid>
  </UnconnectedDatasetsPage>
</Connect(UnconnectedDatasetsPage)>
`;

exports[`DatasetsPage Should match existing snapshot when no props are provided. 1`] = `
<Grid
  centered={true}
  padded={true}
  relaxed={true}
/>
`;

exports[`DatasetsPage Should render dynamics links correctly. 1`] = `
<Grid
  centered={true}
  padded={true}
  relaxed={true}
>
  <GridRow
    columns={3}
  >
    <GridColumn
      textAlign="left"
    >
      <Header />
      <p>
        <span
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Analyses: 2
        </span>
      </p>
      <p>
        <span
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Derived from: 
          Nothing
        </span>
      </p>
      <br />
    </GridColumn>
    <GridColumn>
      <Button
        as="button"
        basic={true}
      >
        <Link
          to="/dynamics?id=some-kind-of-musical-reference&viz=bbfacade-0000-0000-0001-a1234567890b&viz=bbfacade-0000-0000-0003-a1234567890b"
        >
          launch
        </Link>
      </Button>
    </GridColumn>
  </GridRow>
</Grid>
`;
